tensorflow/compiler/jit/deadness_analysis.cc

```diff
@@ -1069,6 +1250,53 @@ Status DeadnessAnalysisImpl::PopulateWithReversePostOrder(
         absl::StrJoin(unreachable_nodes, ", "));
   }
 
+  std::vector<Node*> topo;
+  TF_RETURN_IF_ERROR(
+      GetFrameBasedTopologicalOrder(&graph_, control_flow_info_, &topo));
+
+  size_t frame_start = 0;
+  while (frame_start < topo.size()) {
+    // Batching nodes who have the same root frame.
+    absl::string_view cur_frame_name;
+    TF_RETURN_IF_ERROR(
+        GetRootFrame(topo[frame_start], control_flow_info_, &cur_frame_name));
+    size_t frame_end = frame_start;
+    for (size_t i = frame_start + 1; i < topo.size(); ++i) {
+      absl::string_view i_frame_name;
+      TF_RETURN_IF_ERROR(
+          GetRootFrame(topo[i], control_flow_info_, &i_frame_name));
+      if (i_frame_name == cur_frame_name) {
+        frame_end = i;
+      } else {
+        break;
+      }
+    }
+    absl::Span<Node*> sub_topo(topo.data() + frame_start,
+                               /*length=*/frame_end - frame_start + 1);
+    frame_start = frame_end + 1;
+
+    // First, try the optimistic mode.
+    bool success = false;
+    if (!force_pessimistic && !cur_frame_name.empty()) {
+      TF_RETURN_IF_ERROR(
+          PopulateFrame(sub_topo, /*use_optimistic_mode=*/true, &success));
+    }
+    if (!success) {
+      // The optimistic mode does not converge. Let's fall back to the
+      // pessimistic mode.
+      TF_RETURN_IF_ERROR(
+          PopulateFrame(sub_topo, /*use_optimistic_mode=*/false, nullptr));
+    }
+    VLOG(2) << "Done populating frame " << cur_frame_name << " using the "
+            << (success ? "optimistic" : "pessimistic") << " mode.";
+  }
+
+  return Status::OK();
+}
+
+Status DeadnessAnalysisImpl::PopulateFrame(absl::Span<Node* const> topo,
+                                           bool use_optimistic_mode,
```

Comment:
Let's `CHECK` that `success` is not null if and only if `use_optimistic_mode` is true.  Alternately (I think this is slightly better) if the code guarded under `use_optimistic_mode` is extracted out into a function then we can have the caller do the "was optimism successful" check, letting us remove the `bool *success` out parameter altogether.

Reply:
I chose the former suggestion, i.e., adding a `CHECK`.

I don't see a clear way to implement the later suggestion, as `success` is needed by `Populate()`. I thought about another way to remove `success` by moving the logic about running optimistic and then pessimistic modes from `Populate()` into `PopulateFrame()`. But this makes the `PopulateFrame` more complicated. So, I chose to just implement the `CHECK`.

Reply:
It is probably easier to show what I had in mind (if it works!) by sending a subsequent PR after this is checked in, so I'll do that. :)

Reply:
Sounds fair. :)